lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -648,10 +654,15 @@ def autoscale_view(self, tight=None, scalex=True, scaley=True,
         # This method looks at the rectangular volume (see above)
         # of data and decides how to scale the view portal to fit it.
         if tight is None:
-            # if image data only just use the datalim
-            _tight = self._tight or (
-                len(self.images) > 0
-                and len(self.lines) == len(self.patches) == 0)
+            _tight = self._tight
+            if not _tight:
+                # if image data only just use the datalim
+                for artist in self._children:
+                    if isinstance(artist, mimage.AxesImage):
+                        _tight = True
+                    elif isinstance(artist, (mlines.Line2D, mpatches.Patch)):
+                        _tight = False
+                        break
```

Comment:
```suggestion
            has_image = any(isinstance(c, mimage.AxesImage)
                            for c in self._children)
            has_line_or_patch = any(isinstance(c, (mlines.Line2D, mpatches.Patch))
                                    for c in self._children)
            _tight = self._tight or (has_image and not has_line_or_patch)
```

Reply:
Trying to avoid iterating `self._children` twice here.

Reply:
Why? Optimization? - I don't think that's relevant, 'cause we usually don't have thousands of Artists. And if so, iterating twice through them is probably the least of our problems. Moreover, if performance is a concern, have you profiled this? List comprehensions are better optimized in Python than equivalent for loops, so I wouldn't even know which solution is faster.

Reply:
Given the early exit condition of `any` the performance should be something measured/profiled. In the absence of that, we should go with clarity. It's not clear to me which of those that is.

Reply:
If I've done this correctly, I'm pretty certain that the single loop is faster than the comprehensions, whether the first condition exits early, the second does, or neither:
```python
import timeit


class AxesImage:
    pass


class Line2D:
    pass


class Patch:
    pass


class Other:
    pass


def loop(_tight, children):
    if not _tight:
        # if image data only just use the datalim
        for artist in children:
            if isinstance(artist, AxesImage):
                _tight = True
            elif isinstance(artist, (Line2D, Patch)):
                _tight = False
                break
    return _tight


def comprehension(_tight, children):
    has_image = any(isinstance(c, AxesImage) for c in children)
    has_line_or_patch = any(isinstance(c, (Line2D, Patch))
                            for c in children)
    tight = _tight or (has_image and not has_line_or_patch)
    return tight


setup = 'from __main__ import AxesImage, Line2D, Patch, Other, children'

for N in [10, 100, 1000]:
    print('N =', N)
    children = [Other()] * N
    print('No early exit')
    print('loop',
          timeit.timeit('loop(False, children)', setup=f'{setup}, loop',
                        number=1000000))
    print('comprehension',
          timeit.timeit('comprehension(False, children)',
                        setup=f'{setup}, comprehension', number=1000000))

    children = [Other()] * N
    children[1] = AxesImage()
    print('First loop early exit')
    print('loop',
          timeit.timeit('loop(False, children)', setup=f'{setup}, loop',
                        number=1000000))
    print('comprehension',
          timeit.timeit('comprehension(False, children)',
                        setup=f'{setup}, comprehension', number=1000000))

    children = [Other()] * N
    children[-2] = AxesImage()
    print('First loop late exit')
    print('loop',
          timeit.timeit('loop(False, children)', setup=f'{setup}, loop',
                        number=1000000))
    print('comprehension',
          timeit.timeit('comprehension(False, children)',
                        setup=f'{setup}, comprehension', number=1000000))

    children = [Other()] * N
    children[1] = Line2D()
    print('Second loop early exit')
    print('loop',
          timeit.timeit('loop(False, children)', setup=f'{setup}, loop',
                        number=1000000))
    print('comprehension',
          timeit.timeit('comprehension(False, children)',
                        setup=f'{setup}, comprehension', number=1000000))

    children = [Other()] * N
    children[-2] = Line2D()
    print('Second loop late exit')
    print('loop',
          timeit.timeit('loop(False, children)', setup=f'{setup}, loop',
                        number=1000000))
    print('comprehension',
          timeit.timeit('comprehension(False, children)',
                        setup=f'{setup}, comprehension', number=1000000))

    print()
    print()
```
![Figure_1](https://user-images.githubusercontent.com/302469/96327424-72192c80-1007-11eb-95c5-3417f2b926d8.png)
I don't expect people to have 1000 artists, but the order of 10-100 _might_ be possible in 3D.